src/App.tsx

```diff
@@ -17,9 +17,32 @@ export const App: React.FC = () => {
 
   const [loadedGoodsGroupText, setLoadedGoodsGroupText] = useState('');
 
+  const prevGetGoodsFromServer = useRef<() => Promise<Good[]>>();
+  const shouldSkipLoading = useCallback(
+    (newGetGoodsFromServer: () => Promise<Good[]>) => {
+      if (
+        prevGetGoodsFromServer.current === newGetGoodsFromServer
+        && !hasLoadingError
+      ) {
+        return true;
+      }
+
+      prevGetGoodsFromServer.current = newGetGoodsFromServer;
+
+      return false;
+    },
+    [hasLoadingError],
+  );
```

Comment:
Instead of having this monstrous logic, you can create an enum for button types and store `currentGoodsType` in the state. Then pass it to `handleGoodsLoad` as the second argument and check.